relayer/generic_relayer/pricing/src/utils/currentPricing.ts

```diff
@@ -0,0 +1,229 @@
+};
+
+export async function pullAllCurrentPricingStates(
+  ctx: PricingContext
+): Promise<DeliveryProviderContractState[]> {
+  const allChains = getAllChains(ctx);
+  const states: DeliveryProviderContractState[] = [];
+  for (const chain of allChains) {
+    const state = await pullCurrentPricingState(ctx, chain);
+    states.push(state);
+  }
+  return states;
+}
+
+//Retrieves the current prices from a given chain
+export async function pullCurrentPricingState(
+  ctx: PricingContext,
+  chain: ChainId
+): Promise<DeliveryProviderContractState> {
+  const deliveryProviderAddress = ctx.deliveryProviders[chain as EVMChainId]; //Cast to EVM chain ID type should be safe
+  const DeliveryProvider = DeliveryProvider__factory.connect(
+    deliveryProviderAddress,
+    getEthersProvider(ctx, chain)
+  );
+
+  const state = readState(ctx, chain, DeliveryProvider, getAllChains(ctx));
+  if (state == null) {
+    throw new Error("Could not read state for chain " + chain);
+  } else {
+    //typscript linter is complaining about this line, but it should be safe
+    //@ts-ignore
+    return state;
+  }
+}
+
+//This code is very similar to that in ethereum/ts-scripts/relayer/relayProvider/readRelayProviderContractState.ts,
+//It should be considered for a furture refactor
+
+async function readState(
+  ctx: PricingContext,
+  currentChain: ChainId,
+  deliveryProvider: DeliveryProvider,
+  allChains: ChainId[]
+): Promise<DeliveryProviderContractState | null> {
+  ctx.logger.info(
+    "Gathering relay provider contract status for chain " + currentChain
+  );
+
+  try {
+    const rewardAddress = await deliveryProvider.getRewardAddress();
+    const supportedChains: {
+      chainId: number;
+      isSupported: boolean;
+    }[] = [];
+    const targetChainAddresses: {
+      chainId: number;
+      whAddress: string;
+    }[] = [];
+    const deliveryOverheads: {
+      chainId: number;
+      deliveryOverhead: BigNumber;
+    }[] = [];
+    const maximumBudgets: { chainId: number; maximumBudget: BigNumber }[] = [];
+    const gasPrices: { chainId: number; gasPrice: BigNumber }[] = [];
+    const usdPrices: { chainId: number; usdPrice: BigNumber }[] = [];
+    const assetConversionBuffers: {
+      chainId: number;
+      tolerance: number;
+      toleranceDenominator: number;
+    }[] = [];
+    const owner: string = await deliveryProvider.owner();
+
+    for (const chain of allChains) {
+      supportedChains.push({
+        chainId: chain,
+        isSupported: await deliveryProvider.isChainSupported(chain),
+      });
+
+      targetChainAddresses.push({
+        chainId: chain,
+        whAddress: await deliveryProvider.getTargetChainAddress(chain),
+      });
+
+      deliveryOverheads.push({
+        chainId: chain,
+        deliveryOverhead: await deliveryProvider.quoteDeliveryOverhead(chain),
+      });
+      maximumBudgets.push({
+        chainId: chain,
+        maximumBudget: await deliveryProvider.maximumBudget(chain),
```

Comment:
Is this a xLabs specific pricing process? or a generic one
if the latter, I don't think we can expect that the delivery provider exposes (maximumBudget, nativeCurrencyPrice, assetConversionBuffer, etc) right

Reply:
This is meant for anyone who intends to use the reference DeliveryProvider contracts